enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,302 @@
+the `kubernetes` service, injected through the environment variables. This IP
+is the service network IP, which later on gets translated to an endpoint:
+```bash
+$ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get svc kubernetes -o yaml
+apiVersion: v1
+kind: Service
+metadata:
+  creationTimestamp: "2022-12-20T15:57:23Z"
+  labels:
+    component: apiserver
+    provider: kubernetes
+  name: kubernetes
+  namespace: default
+  resourceVersion: "199"
+  uid: 28e299a6-1578-4fd0-babd-660a1a86cdfb
+spec:
+  clusterIP: 10.43.0.1
+  clusterIPs:
+  - 10.43.0.1
+  internalTrafficPolicy: Cluster
+  ipFamilies:
+  - IPv4
+  ipFamilyPolicy: SingleStack
+  ports:
+  - name: https
+    port: 443
+    protocol: TCP
+    targetPort: 6443
+  sessionAffinity: None
+  type: ClusterIP
+status:
+  loadBalancer: {}
+
+$ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get endpoints kubernetes -o yaml
+apiVersion: v1
+kind: Endpoints
+metadata:
+  creationTimestamp: "2022-12-20T15:57:23Z"
+  labels:
+    endpointslice.kubernetes.io/skip-mirror: "true"
+  name: kubernetes
+  namespace: default
+  resourceVersion: "201"
+  uid: 2c3d37b8-9e04-4779-9947-6230450e2336
+subsets:
+- addresses:
+  - ip: 192.168.122.117
+  ports:
+  - name: https
+    port: 6443
+    protocol: TCP
+```
+As we can see the endpoint is the node IP, which means that any connection to
+the service IP will get translated to the node IP. When reaching API server the
+destination IP will be that of the node, and if there is no match for that in
+the map the default certificate is returned. This works because the service
+network certificate includes the service IP for hostname validation.
+
+#### Including the IP in the certificates
+We have already seen that internal access always uses the node IP because of
+the endpoints.
+
+If we were to include the node IP in the service network certificate we would
+get an immediate match in the mapping for pods, and the hostname verification
+matches the service network IP. If the connection using the node IP as
+destination does not come from a pod we run into trouble. The service network
+certificate is signed by its own specific CA, so any external connection using
+the IP will get the internal certificate and wont validate against the external
+CA.
+
+If we were to include the node IP in the external certificate we also get an
+immediate match in the mapping, and the hostname verification also works ok.
+All internal connections (those coming from pods through the service network
+IP) will be broken when trying to validate hostname, as the service IP is
+not included. Even if it was included, validation with the internal CA would
+still fail.
+
+Including the IP in the current situation means choosing what to lose:
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity.
```

Comment:
You could say here something like "MicroShift will run in environments where hostname access may not be guaranteed. Customers have given us the requirement of accessing the cluster via IP address."